### ISSUE TITLE
Creating Jobs with the same ID has unexpected behavior

### ISSUE THREAD

**Issue by nw-ct** [rq/rq#1701]
### Expected behavior:

- Enqueuing two jobs with the same ID on the same queue should only keep the later one
- Enqueuing a job in two different queues with the same job ID should enqueue both jobs

### Actual behavior:

- Enqueueing two jobs with the same ID on the same queue keeps two jobs, but the data on the first job is overwritten by the data on the second job.
- Enqueuing a job in two different queues with the same job ID deletes the first job, leaving only the job on the second queue

#### Same queue (deletes first job, duplicates second job):

```python
redis_connection.flushall()
low = rq.Queue("my_test_low", connection=redis_connection)
high = rq.Queue("my_test_high", connection=redis_connection)
j = low.enqueue_call(f, job_id="my_job_id", args=("a",))
sleep(1)
k = low.enqueue_call(f, job_id="my_job_id", args=("b",))
print([(x.id, x.enqueued_at, x.args) for x in low.jobs])
print([(x.id, x.enqueued_at, x.args) for x in high.jobs])
```

```
[('my_job_id', datetime.datetime(2022, 8, 19, 14, 13, 10, 933287), ('b',)), ('my_job_id', datetime.datetime(2022, 8, 19, 14, 13, 10, 933287), ('b',))]
[]
```
Notice how the information in both jobs is the same ('b' is the argument, and the timestamp matches despite the sleep)

#### Different queue (deletes first job, keeps only second job):
```python
redis_connection.flushall()
low = rq.Queue("my_test_low", connection=redis_connection)
high = rq.Queue("my_test_high", connection=redis_connection)
j = low.enqueue_call(f, job_id="my_job_id", args=("a",))
sleep(1)
k = high.enqueue_call(f, job_id="my_job_id", args=("b",))
print([(x.id, x.enqueued_at, x.args) for x in low.jobs])
print([(x.id, x.enqueued_at, x.args) for x in high.jobs])
```

```
[]
[('my_job_id', datetime.datetime(2022, 8, 19, 14, 14, 31, 65815), ('b',))]
```
Notice how the first job was deleted, and only the second job remains, despite being completely different queues.

**Comment by selwin** [comment]
Yeah, I think we can try prefixing the job ID with the name of the queue. But I'm not sure if this will be backwards compatible.

Alternatively, I think we can add a note on the docs that job IDs have to be unique globally.

Related to this I also want to implement duplicate detection based on job ID.

**Comment by nw-ct** [comment]
Thanks for the response selwin!

Prefixing the job_id with the name of the queue would fix the situation where the queues are different, but there would still be an issue in the situation where the queues are the same, right?

Documenting something like: "Job IDs should be globally unique - enqueueing two jobs with the same ID has unpredictable behavior" would help devs avoid this situation in the first place. Preferable though, there should still be consistent semantics for what should happen if a dev does it anyway. Right now it's different based on same-queue different-queue.

It would be nice if the system could detect if a duplicate is being enqueued, and either delete the old one or reject the new one. Is that a heavy lift?

**Comment by nw-ct** [comment]
Friendly ping @selwin :)

**Comment by selwin** [comment]
Hey @nw-ct thanks for the reminder.

Duplicate job detection is something I've always been meaning to develop but my immediate goal is to land a feature that lets RQ saves multiple execution results here: https://github.com/rq/rq/pull/1573

> It would be nice if the system could detect if a duplicate is being enqueued, and either delete the old one or reject the new one. Is that a heavy lift?
I outlined my proposed solution for duplicate detection here, it would be great if someone is willing to take a crack at it: https://github.com/rq/rq/issues/1716#issuecomment-1278292467

**Comment by nw-ct** [comment]
Hi @selwin! It looks like the feature to save multiple execution results landed (nice!), so I wanted to follow up on job deduplication. Any chance this might be worked on in the near future?

**Comment by selwin** [comment]
Yes, this might be next on my priority list.
